database/src/main/java/com/firebase/ui/database/FirebaseArray.java

```diff
@@ -14,52 +14,118 @@
 
 package com.firebase.ui.database;
 
+import android.support.annotation.NonNull;
+
 import com.google.firebase.database.ChildEventListener;
 import com.google.firebase.database.DataSnapshot;
 import com.google.firebase.database.DatabaseError;
 import com.google.firebase.database.Query;
 import com.google.firebase.database.ValueEventListener;
 
 import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
 import java.util.List;
+import java.util.ListIterator;
 
 /**
- * This class implements an array-like collection on top of a Firebase location.
+ * This class implements a collection on top of a Firebase location.
  */
-class FirebaseArray implements ChildEventListener, ValueEventListener {
+public class FirebaseArray extends ImmutableList<DataSnapshot> implements ChildEventListener, ValueEventListener {
     private Query mQuery;
-    private ChangeEventListener mListener;
+    private boolean mNotifyListeners = true;
+    private List<ChangeEventListener> mListeners = new ArrayList<>();
+    private List<SubscriptionEventListener> mSubscribers = new ArrayList<>();
     private List<DataSnapshot> mSnapshots = new ArrayList<>();
 
-    public FirebaseArray(Query ref) {
-        mQuery = ref;
-        mQuery.addChildEventListener(this);
-        mQuery.addValueEventListener(this);
+    /**
+     * @param query The Firebase location to watch for data changes. Can also be a slice of a
+     *              location, using some combination of {@code limit()}, {@code startAt()}, and
+     *              {@code endAt()}.
+     */
+    public FirebaseArray(Query query) {
+        mQuery = query;
     }
 
-    public void cleanup() {
-        mQuery.removeEventListener((ValueEventListener) this);
-        mQuery.removeEventListener((ChildEventListener) this);
+    /**
+     * Add a listener for change events and errors occurring at the location provided in {@link
+     * #FirebaseArray(Query)}.
+     *
+     * @param listener the listener to be called with changes
+     * @return a reference to the listener provided. Save this to remove the listener later
+     * @throws IllegalArgumentException if the listener is null
+     */
+    public ChangeEventListener addChangeEventListener(@NonNull ChangeEventListener listener) {
+        checkNotNull(listener);
+
+        mListeners.add(listener);
+        notifySubscriptionEventListeners(SubscriptionEventListener.EventType.ADDED);
+        if (mListeners.size() <= 1) {
```

Comment:
nit: change to `mListeners.size() == 1` and add a comment explaining this pattern.  As I understand it, we are not listening to Firebase Database until at least one listener has been added.